File: app/src/main/java/com/happs/ximand/ringcontrol/viewmodel/fragment/ExceptionViewModel.java
package com.happs.ximand.ringcontrol.viewmodel.fragment;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import com.happs.ximand.ringcontrol.SingleLiveEvent;
import com.happs.ximand.ringcontrol.model.object.exception.bl.BluetoothException;

public class ExceptionViewModel extends BaseViewModel {

    private final SingleLiveEvent<Void> restartApplicationLiveEvent;
    private MutableLiveData<BluetoothException> exceptionLiveData = new MutableLiveData<>();

    public ExceptionViewModel() {
        this.restartApplicationLiveEvent = new SingleLiveEvent<>();
    }

    public SingleLiveEvent<Void> getRestartApplicationLiveEvent() {
        return restartApplicationLiveEvent;
    }

    public LiveData<BluetoothException> getExceptionLiveData() {
        return exceptionLiveData;
    }

    public void setException(BluetoothException exception) {
        exceptionLiveData.setValue(exception);
    }

    public void onRetryClick() {
        restartApplicationLiveEvent.call();
    }
}

File: app/src/main/java/com/happs/ximand/ringcontrol/viewmodel/fragment/ExceptionViewModel.kt
package com.happs.ximand.ringcontrol.viewmodel.fragment

import com.happs.ximand.ringcontrol.R
import com.happs.ximand.ringcontrol.SingleLiveEvent
import com.happs.ximand.ringcontrol.model.`object`.exception.BaseException
import com.happs.ximand.ringcontrol.model.`object`.exception.NoOneEmailClientException

class ExceptionViewModel : BaseViewModel() {

    companion object {
        const val SUPPORT_EMAIL = "ximand931@gmail.com"
    }

    val navigateToPermissionSettingsLiveEvent = SingleLiveEvent<Void>()
    val navigateToSupportLiveEvent = SingleLiveEvent<String>()
    val restartApplicationLiveEvent = SingleLiveEvent<Void>()

    var exception: BaseException? = null

    fun attachException(exception: BaseException) {
        this.exception = exception
    }

    fun restartApplication() {
        restartApplicationLiveEvent.call()
    }

    fun navigateToPermissionsSettings() {
        navigateToPermissionSettingsLiveEvent.call()
    }

    fun navigateToSupport() {
        navigateToSupportLiveEvent.value = SUPPORT_EMAIL
    }

    fun onNoOneEmailClient() {
        makeExceptionSnackbarWithAction(R.string.no_one_client_title, NoOneEmailClientException())
    }

}